charts/secrets-store-sync-controller/templates/_helpers.tpl

```diff
@@ -0,0 +1,95 @@
+{{- end -}}
+
+{{/*
+Generate allowed secret types list as a complete expression.
+*/}}
+{{- define "chartname.secretTypesList" -}}
+{{- $secretTypes := . -}}
+{{- if not $secretTypes -}}
+false
+{{- else -}}
+(object.type in [{{ range $index, $type := $secretTypes }}{{ if $index }}, {{ end }}"{{ $type }}"{{ end }}])
+{{- end -}}
+{{- end -}}
+
+{{/* Define a constant value for labelKey */}}
+{{- define "secrets-store-sync-controller.labelKey" -}}
+secrets-store.sync.x-k8s.io
+{{- end -}}
+
+{{/* Define a constant value for labelValue */}}
+{{- define "secrets-store-sync-controller.labelValue" -}}
+''
+{{- end -}}
+
+{{/*
+Check if the old secret has the expected label key.
+*/}}
+{{- define "secrets-store-sync-controller.oldSecretHasExpectedLabelKey" -}}
+variables.oldSecretHasLabels && ('{{ include "secrets-store-sync-controller.labelKey" . }}' in oldObject.metadata.labels) ? true : false
+{{- end -}}
+
+{{/*
+Check if the old secret has the expected label value.
+*/}}
+{{- define "secrets-store-sync-controller.oldSecretHasExpectedLabelValue" -}}
+{{ include "secrets-store-sync-controller.labelValue" . }} == oldObject.metadata.labels['{{ include "secrets-store-sync-controller.labelKey" . }}'] ? true : false
+{{- end -}}
+
+
+{{/*
+Generate token audience comparison expression.
+Returns 'false' if tokenRequestAudience list is empty.
+*/}}
+{{- define "secrets-store-sync-controller.tokenAudienceComparison" -}}
+{{- $tokenAudiences := .Values.tokenRequestAudience -}}
+{{- if not $tokenAudiences -}}
+false
+{{- else -}}
+{{- $audienceExpressions := list -}}
+{{- range $index, $audience := $tokenAudiences }}
+  {{- $expressionPart := printf "object.spec.audiences.exists(w, w == '%s')" $audience.audience -}}
+  {{- $audienceExpressions = append $audienceExpressions $expressionPart -}}
+{{- end -}}
+{{- join " || " $audienceExpressions -}}
+{{- end -}}
+{{- end -}}
+
+{{/*
+Generate a comma-separated string from a list.
+*/}}
+{{- define "secrets-store-sync-controller.listToString" -}}
+{{- $tokenRequests := .Values.tokenRequestAudience -}}
+{{- $audiences := list -}}
+{{- range $index, $request := $tokenRequests }}
+  {{- $audiences = append $audiences $request.audience -}}
+{{- end -}}
+{{- join ", " $audiences -}}
+{{- end -}}
+
+{{/*
+Determine the api version for the validating admission policies.
+*/}}
+{{- define "secrets-store-sync-controller.admissionApiVersion" -}}
+{{- if semverCompare "~1.27.0" .Values.validatingAdmissionPolicies.kubernetesReleaseVersion -}}
+apiVersion: admissionregistration.k8s.io/v1alpha1
+{{- else if semverCompare "~1.28.0" .Values.validatingAdmissionPolicies.kubernetesReleaseVersion -}}
+apiVersion: admissionregistration.k8s.io/v1beta1
+{{- else if semverCompare "~1.29.0" .Values.validatingAdmissionPolicies.kubernetesReleaseVersion -}}
+apiVersion: admissionregistration.k8s.io/v1
+{{- else -}}
+apiVersion: unsupported-validating-admission-api-version
+{{- end }}
+{{- end -}}
```

Comment:
Wouldn't this check fail for 1.30+? According to https://helm.sh/docs/chart_template_guide/function_list/#semvercompare 

```
semverCompare "~1.2.0" "1.2.3" checks that the major and minor versions match, and that the patch number of the second version is greater than or equal to the first parameter.
```

Reply:
I tried with v1.31.0 and it works

```shell
$ KIND_NODE_IMAGE_VERSION=v1.31.0 make local-setup                   

...

kind create cluster --name sync-controller \
                --image kindest/node:v1.31.0 \
                --config=./hack/localsetup/kind-config.yaml
Creating cluster "sync-controller" ...
 ✓ Ensuring node image (kindest/node:v1.31.0) 🖼 
 ✓ Preparing nodes 📦  
 ✓ Writing configuration 📜 
 ✓ Starting control-plane 🕹️ 
 ✓ Installing CNI 🔌 
 ✓ Installing StorageClass 💾 
Set kubectl context to "kind-sync-controller"
You can now use your cluster with:

kubectl cluster-info --context kind-sync-controller

Not sure what to do next? 😅  Check out https://kind.sigs.k8s.io/docs/user/quick-start/

...

helm install secrets-store-sync-controller --wait --timeout=5m \
                -f manifest_staging/charts/secrets-store-sync-controller/temp_values.yaml \
                --set image.repository=docker.io/controller \
                --set image.tag=v0.0.1 \
                manifest_staging/charts/secrets-store-sync-controller

...

NAME: secrets-store-sync-controller
LAST DEPLOYED: Tue Aug 20 16:36:54 2024
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE: None
```

API Version is correctly set in VAP

```shell
$ k get ValidatingAdmissionPolicy secrets-store-sync-controller-create-update-policy
NAME                                                 VALIDATIONS   PARAMKIND   AGE
secrets-store-sync-controller-create-update-policy   1             <unset>     9m19s
$ k get ValidatingAdmissionPolicy secrets-store-sync-controller-create-update-policy -oyaml
apiVersion: admissionregistration.k8s.io/v1
kind: ValidatingAdmissionPolicy
metadata:
  annotations:
    meta.helm.sh/release-name: secrets-store-sync-controller
    meta.helm.sh/release-namespace: default
  creationTimestamp: "2024-08-20T23:36:54Z"
  generation: 1
  labels:
    app.kubernetes.io/managed-by: Helm
  name: secrets-store-sync-controller-create-update-policy
  resourceVersion: "343"
  uid: 77679c80-271a-4edc-928e-7908ea5ca336
spec:
  failurePolicy: Fail
  matchConditions:
  - expression: request.userInfo.username == 'system:serviceaccount:secrets-store-sync-controller-system:secrets-store-sync-controller-manager'
    name: user-is-secrets-store-sync-controller
  matchConstraints:
    matchPolicy: Equivalent
    namespaceSelector: {}
    objectSelector: {}
    resourceRules:
    - apiGroups:
      - ""
      apiVersions:
      - v1
      operations:
      - CREATE
      - UPDATE
      resources:
      - secrets
      scope: '*'
  validations:
  - expression: variables.allowedSecretTypes == true && variables.hasOneSecretSyncOwner
      == true
    message: Only secrets with types defined in the allowedSecretTypes are allowed.
    messageExpression: '''secrets-store-sync-controller has failed to '' +  string(request.operation)
      + '' secret with '' + string(object.type) + '' type '' + ''in the '' + string(request.namespace)
      + '' namespace. The controller can only create or update secrets in the allowed
      types list with a single secretsync owner.'''
  variables:
  - expression: (size(object.metadata.ownerReferences) == 1 && object.metadata.ownerReferences.all(o,
      o.kind == 'SecretSync' && o.apiVersion.startsWith('secret-sync.x-k8s.io/') &&
      o.name == object.metadata.name))
    name: hasOneSecretSyncOwner
  - expression: (object.type in ["Opaque", "kubernetes.io/basic-auth", "bootstrap.kubernetes.io/token",
      "kubernetes.io/dockerconfigjson", "kubernetes.io/dockercfg", "kubernetes.io/ssh-auth",
      "kubernetes.io/tls"])
    name: allowedSecretTypes
status:
  observedGeneration: 1
  typeChecking: {}
```

Reply:
This doesn't actually test the helm function. You need to set `validatingAdmissionPolicies.kubernetesReleaseVersion` to test if it works.

The command fails as per my comment (https://github.com/kubernetes-sigs/secrets-store-sync-controller/pull/58#discussion_r1723954788)

```bash
➜ helm template secrets-store-sync-controller --debug --set validatingAdmissionPolicies.kubernetesReleaseVersion="1.31.0" \
                --set image.repository=docker.io/controller \
                --set image.tag=v0.0.1 \
                manifest_staging/charts/secrets-store-sync-controller | grep apiVersion
install.go:222: [debug] Original chart version: ""
install.go:239: [debug] CHART PATH: /home/anramase/go/src/sigs.k8s.io/secrets-store-sync-controller/manifest_staging/charts/secrets-store-sync-controller

apiVersion: v1
apiVersion: v1
apiVersion: rbac.authorization.k8s.io/v1
apiVersion: rbac.authorization.k8s.io/v1
apiVersion: rbac.authorization.k8s.io/v1
apiVersion: rbac.authorization.k8s.io/v1
apiVersion: rbac.authorization.k8s.io/v1
apiVersion: rbac.authorization.k8s.io/v1
apiVersion: apps/v1
apiVersion: unsupported-validating-admission-api-version
        apiVersions: ["v1"]
      expression: "(size(object.metadata.ownerReferences) == 1 && object.metadata.ownerReferences.all(o, o.kind == 'SecretSync' && o.apiVersion.startsWith('secret-sync.x-k8s.io/') && o.name == object.metadata.name))"
apiVersion: unsupported-validating-admission-api-version
        apiVersions: ["v1"]
apiVersion: unsupported-validating-admission-api-version
      apiVersions: ["v1"]
apiVersion: unsupported-validating-admission-api-version
    apiVersion: v1
      apiVersions: ["v1"]
    expression: "has(oldObject.metadata.ownerReferences) && (size(oldObject.metadata.ownerReferences) == 1 && oldObject.metadata.ownerReferences.all(o, o.kind == 'SecretSync' && o.apiVersion.startsWith('secret-sync.x-k8s.io/') && o.name == object.metadata.name))"
apiVersion: unsupported-validating-admission-api-version
      apiVersions: ["v1"]
apiVersion: unsupported-validating-admission-api-version
      apiVersions: ["*"]
apiVersion: unsupported-validating-admission-api-version
      apiVersions: ["*"]
apiVersion: unsupported-validating-admission-api-version
      apiVersions: ["v1"]
apiVersion: unsupported-validating-admission-api-version
apiVersion: unsupported-validating-admission-api-version
apiVersion: unsupported-validating-admission-api-version
apiVersion: unsupported-validating-admission-api-version
apiVersion: unsupported-validating-admission-api-version
apiVersion: unsupported-validating-admission-api-version
apiVersion: unsupported-validating-admission-api-version
apiVersion: unsupported-validating-admission-api-version
```

Reply:
I made the changes in semverCompare and it now works

```shell
$ KIND_NODE_IMAGE_VERSION=v1.31.0 make local-setup
...

kind create cluster --name sync-controller \
                --image kindest/node:v1.31.0 \
                --config=./hack/localsetup/kind-config.yaml
Creating cluster "sync-controller" ...
 ✓ Ensuring node image (kindest/node:v1.31.0) 🖼 
 ✓ Preparing nodes 📦  
 ✓ Writing configuration 📜 
 ✓ Starting control-plane 🕹️ 
 ✓ Installing CNI 🔌 
 ✓ Installing StorageClass 💾 
Set kubectl context to "kind-sync-controller"
You can now use your cluster with:

kubectl cluster-info --context kind-sync-controller

Not sure what to do next? 😅  Check out https://kind.sigs.k8s.io/docs/user/quick-start/

...

helm install secrets-store-sync-controller --wait --timeout=5m \
                --namespace secrets-store-sync-controller-system --create-namespace \
                -f manifest_staging/charts/secrets-store-sync-controller/temp_values.yaml \
                --set image.repository=docker.io/controller \
                --set image.tag=v0.0.1 \
                --set validatingAdmissionPolicies.kubernetesReleaseVersion="1.31.0" \
                manifest_staging/charts/secrets-store-sync-controller
NAME: secrets-store-sync-controller
LAST DEPLOYED: Wed Aug 21 10:49:39 2024
NAMESPACE: secrets-store-sync-controller-system
STATUS: deployed
REVISION: 1
TEST SUITE: None


$ k get validatingAdmissionPolicies secrets-store-sync-controller-create-update-policy -oyaml
apiVersion: admissionregistration.k8s.io/v1
kind: ValidatingAdmissionPolicy
metadata:
  annotations:
    meta.helm.sh/release-name: secrets-store-sync-controller
    meta.helm.sh/release-namespace: secrets-store-sync-controller-system
  creationTimestamp: "2024-08-21T17:49:40Z"
  generation: 1
  labels:
    app.kubernetes.io/managed-by: Helm
  name: secrets-store-sync-controller-create-update-policy
  resourceVersion: "414"
  uid: b0529446-83c6-4498-bdec-1def53c11e24
spec:
  failurePolicy: Fail
  matchConditions:
  - expression: request.userInfo.username == 'system:serviceaccount:secrets-store-sync-controller-system:secrets-store-sync-controller-manager'
    name: user-is-secrets-store-sync-controller
  matchConstraints:
    matchPolicy: Equivalent
    namespaceSelector: {}
    objectSelector: {}
    resourceRules:
    - apiGroups:
      - ""
      apiVersions:
      - v1
      operations:
      - CREATE
      - UPDATE
      resources:
      - secrets
      scope: '*'
  validations:
  - expression: variables.allowedSecretTypes == true && variables.hasOneSecretSyncOwner
      == true
    message: Only secrets with types defined in the allowedSecretTypes are allowed.
    messageExpression: '''secrets-store-sync-controller has failed to '' +  string(request.operation)
      + '' secret with '' + string(object.type) + '' type '' + ''in the '' + string(request.namespace)
      + '' namespace. The controller can only create or update secrets in the allowed
      types list with a single secretsync owner.'''
  variables:
  - expression: (size(object.metadata.ownerReferences) == 1 && object.metadata.ownerReferences.all(o,
      o.kind == 'SecretSync' && o.apiVersion.startsWith('secret-sync.x-k8s.io/') &&
      o.name == object.metadata.name))
    name: hasOneSecretSyncOwner
  - expression: (object.type in ["Opaque", "kubernetes.io/basic-auth", "bootstrap.kubernetes.io/token",
      "kubernetes.io/dockerconfigjson", "kubernetes.io/dockercfg", "kubernetes.io/ssh-auth",
      "kubernetes.io/tls"])
    name: allowedSecretTypes
status:
  observedGeneration: 1
  typeChecking: {}

```